### PR TITLE
TCP: Connect Timeout; Close Stream

### PR DESCRIPTION
**Pull request by garyrussell**
- Add `connectTimeout` to client connection factories
- Add `closeStreamAfterSend` to outbound gateway

### CODE REVIEW

File: spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpOutboundGateway.java [hunk 1]
```diff
@@ -117,6 +120,8 @@ protected void doInit() {
 		if (!this.evaluationContextSet) {
 			this.evaluationContext = ExpressionUtils.createStandardEvaluationContext(getBeanFactory());
 		}
+		Assert.state(this.closeStreamAfterSend ? this.isSingleUse : true,
```
Comment (artembilan) [inline spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/TcpOutboundGateway.java]: `!this.closeStreamAfterSend || this.isSingleUse` ?

File: spring-integration-ip/src/test/java/org/springframework/integration/ip/dsl/IpIntegrationTests.java [hunk 2]
```diff
@@ -164,6 +172,43 @@ public void testUdpInheritance() {
 		assertThat(udpMulticastOutboundChannelAdapterSpec2.get()).isInstanceOf(MulticastSendingMessageHandler.class);
 	}
 
+	@Test
+	public void testCloseStream() throws InterruptedException {
+		IntegrationFlow server = IntegrationFlows.from(Tcp.inboundGateway(Tcp.netServer(0)
+				.deserializer(new ByteArrayRawSerializer())))
+				.<byte[], String>transform(p -> "reply:" + new String(p).toUpperCase())
+				.bridge()
```
Comment (artembilan) [inline spring-integration-ip/src/test/java/org/springframework/integration/ip/dsl/IpIntegrationTests.java]: I wonder why do we need bridge here...

File: spring-integration-ip/src/test/java/org/springframework/integration/ip/dsl/IpIntegrationTests.java [hunk 3]
```diff
@@ -164,6 +172,43 @@ public void testUdpInheritance() {
...
+			.handle(Tcp.outboundGateway(Tcp.netClient("localhost", port.get())
+					.singleUseConnections(true)
+					.serializer(new ByteArrayRawSerializer()))
+				.closeStreamAfterSend(true))
+			.transform(Transformers.objectToString())
+			.get();
+		this.flowContext.registration(client)
+			.id("streamCloseClient")
+			.register();
+		MessagingTemplate messagingTemplate = new MessagingTemplate(client.getInputChannel());
```
Comment (artembilan) [inline spring-integration-ip/src/test/java/org/springframework/integration/ip/dsl/IpIntegrationTests.java]: You can use `IntegrationFlowRegistration.getMessagingTemplate()` after the `.register()`

File: spring-integration-ip/src/test/java/org/springframework/integration/ip/tcp/TcpInboundGatewayTests.java [hunk 4]
```diff
@@ -284,6 +291,57 @@ public void testErrorFlow() throws Exception {
...
+			return false;
+		});
+		client.afterPropertiesSet();
+		client.start();
+		TcpConnectionSupport connection = client.getConnection();
+		connection.send(new GenericMessage<>("foo"));
+		connection.shutdownOutput(); // signal EOF to server
+		assertThat(latch.await(10, TimeUnit.SECONDS)).isTrue();
+		gateway.stop();
+		client.stop();
```
Comment (artembilan) [inline spring-integration-ip/src/test/java/org/springframework/integration/ip/tcp/TcpInboundGatewayTests.java]: Don't we need to destroy `scf` in the end as well?

**Review by artembilan — CHANGES_REQUESTED** [review]

File: spring-integration-ip/src/test/java/org/springframework/integration/ip/dsl/IpIntegrationTests.java (hunk 2 above)
Comment (garyrussell) [reply spring-integration-ip/src/test/java/org/springframework/integration/ip/dsl/IpIntegrationTests.java]: Good; I added the transformer later and should have removed the bridge.

File: spring-integration-ip/src/test/java/org/springframework/integration/ip/tcp/TcpInboundGatewayTests.java (hunk 4 above)
Comment (garyrussell) [reply spring-integration-ip/src/test/java/org/springframework/integration/ip/tcp/TcpInboundGatewayTests.java]: No; the gateway stops the scf (and it's not a `DisposableBean`).

File: src/reference/asciidoc/whats-new.adoc [hunk 5]
```diff
@@ -69,6 +69,13 @@ See <<./ip.adoc#tcp-codecs,Message Demarcation (Serializers and Deserializers)>>
 When using a `TcpNioServerConnectionFactory`, priority is now given to accepting new connections over reading from existing connections, but it is configurable.
 See <<./ip.adoc#note-nio,About Non-blocking I/O (NIO)>> for more information.
 
+The outbound gateway has a new property `closeStreamAfterSend`; when used with a new connection for each request/reply it signals EOF to the server, without closing the connection.
+This is useful for servers that use the EOF to signal end of message instead of some delimiter in the data.
+See <<tcp-gateways>> for more information.
```
Comment (artembilan) [inline src/reference/asciidoc/whats-new.adoc]: Doh! I forgot that we need to be multi-page ready 😄 : 
```
Execution failed for task ':checkAsciidocLinks'.
> [*** Anchor reference errors found:,
  Anchor 'tcp-gateways' not found in 'whats-new.adoc', if in another file, it needs to be qualified with './fileName.adoc#',
  Anchor 'tcp-connection-factory' not found in 'whats-new.adoc', if in another file, it needs to be qualified with './fileName.adoc#']
```

Can you fix it? 
Thanks

**Pull request merged**